src/tools/illink/src/ILLink.Shared/DataFlow/ITransfer.cs

```diff
@@ -23,16 +23,29 @@ namespace ILLink.Shared.DataFlow
 
 	// TLattice isn't typically used in the implementation except to provide the "Top" value.
 	// This expresses the conceptual constraint that the transferred values are part of a lattice.
-	public interface ITransfer<TOperation, TValue, TState, TLattice>
+	public interface INegate<TValue>
+	{
+		TValue Negate ();
+	}
+
+	public interface ITransfer<TOperation, TValue, TState, TLattice, TConditionValue>
 		where TValue : struct, IEquatable<TValue>
 		where TState : class, IDataFlowState<TValue, TLattice>
 		where TLattice : ILattice<TValue>
+		where TConditionValue : struct, INegate<TConditionValue>
 	{
 		// Transfer should mutate the input value to reflect the effect of
 		// computing this operation. When using value types, ensure that
 		// any modifications to the values are observable by the caller (consider
 		// using readonly structs to prevent the implementation from making changes
 		// that won't be reflected in the caller).
-		void Transfer (TOperation operation, TState state);
+		// The optional return value can be used to model branch conditions.
+		TConditionValue? Transfer (TOperation operation, TState state);
+
+		// Applies a condition value to a dataflow state. This is called in the
+		// outgoing branch of a basic block which returns a TConditionValue
+		// to model the state in the conditional branch where the condition is
+		// assumed to be true.
+		void ApplyCondition (TConditionValue condition, ref TValue value);
```

Comment:
"Applies a condition value to a dataflow state." this is confusing. It applies the condition to a "value" (per the interface terminology).

Aside - the "value" here is super confusing, because in the usage we actually call that thing a "state" as well - but that's not something we should fix with this PR.

Reply:
I'm not ready for that naming discussion :-)